### DISCUSSION TITLE
how to use routes.MapAreaRoute with three parameters?

### DISCUSSION THREAD

**Discussion by mmdqq** [OrchardCMS/OrchardCore#15839] [Q&A]
![image](https://user-images.githubusercontent.com/4881282/58678937-15768c80-8394-11e9-81ec-b76787aa950c.png)
![image](https://user-images.githubusercontent.com/4881282/58680184-3b526000-8399-11e9-986f-a60bf5ed94c2.png)
![image](https://user-images.githubusercontent.com/4881282/58678984-43f46780-8394-11e9-9286-9436f56d0620.png)
The  routes.MapAreaRoute named “SubMenuCreate” didn't work at all.

**Comment by Skrypt** [comment]
Try this please.

```C#
public async Task<IActionResult> Create(string id, string menuContentItemId, string? menuItemId){

}
```

**Comment by mmdqq** [comment]
@Skrypt string? case error compile

**Comment by Skrypt** [comment]
Oh really this doesn't work?

**Comment by mmdqq** [comment]
@Skrypt  yes you can try it,i search all of orchardcore code ,there is none routes.MapAreaRoute use three parameters

**Comment by Skrypt** [comment]
hmmm You are right in fact this doesn't work at all sorry mate my error. Though I don't understand why you have this issue. Look at the latest source and this is all working.

**Comment by mmdqq** [comment]
I want to use simple URLs in a unified way ，OrchardCore.xxx and & in Url is too long

**Comment by Skrypt** [comment]
```C#
public async Task<IActionResult> Create(string id, string menuContentItemId, string menuItemId = ""){

}
```

**Comment by Skrypt** [comment]
I can't find this routes.MapAreaRoute anywhere in OC so I presume this is custom code right?

**Comment by mmdqq** [comment]
![image](https://user-images.githubusercontent.com/4881282/58680104-eb739900-8398-11e9-9aa5-0cadc38fc086.png)
comile is  OK, but the Url not change

**Comment by mmdqq** [comment]
yes ,

```C#
namespace OrchardCore.Menu
{
    public class Startup : StartupBase
    {
        public override void ConfigureServices(IServiceCollection services)
        {
            services.AddScoped<IDataMigration, Migrations>();
            services.AddScoped<IShapeTableProvider, MenuShapes>();
            services.AddScoped<IPermissionProvider, Permissions>();

            // MenuPart
            services.AddScoped<IContentHandler, MenuContentHandler>();
            services.AddScoped<IContentPartDisplayDriver, MenuPartDisplayDriver>();
            services.AddSingleton<ContentPart, MenuPart>();
            services.AddSingleton<ContentPart, MenuItemsListPart>();

            // LinkMenuItemPart
            services.AddScoped<IContentPartDisplayDriver, LinkMenuItemPartDisplayDriver>();
            services.AddSingleton<ContentPart, LinkMenuItemPart>();

            services.AddTagHelpers<MenuTagHelper>();
        }
       public override void Configure(IApplicationBuilder builder, IRouteBuilder routes, IServiceProvider serviceProvider)
        {
            routes.MapAreaRoute(
                name: "SubMenuCreate",
                areaName: "OrchardCore.Menu",
                template: "SubMenu/Create/{id}/{menuContentItemId}/{menuItemId}",
                defaults: new { controller = "Admin", action = "Create" }
            );
        }
    }
}
```

**Comment by Skrypt** [comment]
By the way it's faster just sending code than sending images. You can use github markdown code :  ```

**Comment by mmdqq** [comment]
OK , but how to fix this problem

**Comment by Skrypt** [comment]
I'm looking at the code to understand which route the controller action uses normally.

**Comment by Skrypt** [comment]
Ok so the **Create** action of this controller is for adding a **menu item** to a **Menu**. But when you click on the "add" button that is on a "menu item" it will pass the current menu item clicked as the "menuItemId" that will be his "parent" to add it under it.

![image](https://user-images.githubusercontent.com/3228637/58680820-8f2c5d00-8337-11e9-80ad-f3f495adff33.png)

So what you are trying to do is use the OrchardCore.Menu module outside OC context I guess. Which is why you need to add the **routes.MapAreaRoute()** method in your startup. You should not change the controller params to have a default value then. You should just pass the proper menu item id's to the controller action so that it works properly. If you are saying that the action is never getting hit then the issue is with the route declaration.

**Comment by jtkech** [comment]
Seems to work on my side

![SubMenu](https://user-images.githubusercontent.com/8586360/58681089-2c54b900-836b-11e9-8ade-9a17cffc3f71.png)

**Comment by Skrypt** [comment]
Yes, @jtkech do you know where the routes are added for the menus? Maybe in the abstractions? But here he seems to say that this is not working for him so I'm guessing that he's trying to use the module as a standalone module for something else than OC.

**Comment by Skrypt** [comment]
We previously closed the issue because this is a none issue if we can't repro in OC. Unless @mmdqq sends us repro steps of an issue then this will be tagged as a discussion because it's too vague.

**Comment by Skrypt** [comment]
And I always forget that `string?` doesn't work 😄

**Comment by jtkech** [comment]
Routes are defined globally in `OC.Contents`, for `Create` it uses `{id}` which is the content type.

Just copy paste the above `MapAreaRoute` and it doesn't work, so maybe just a typo, let me check.

**Comment by mmdqq** [comment]
@jtkech  you Url just one parameter,you can try  use @Skrypt 's example to add "324234"'s submenu,this way need three parameters

**Comment by jtkech** [comment]
Okay, the above `MapAreaRoute` was just missing the `?` at the end of `{menuItemId?}`.

As i have seen, `menuItemId` is the parent menu, if null the menu item is added to the actual menu.

Okay, i tried this url `https://localhost:44300/SubMenu/Create/LinkMenuItem/4zv8qc23wtjgx1f8cgm89ehxmt/324234`

Then i got this page.

![SubMenu](https://user-images.githubusercontent.com/8586360/58681820-6f645b80-836e-11e9-816d-147af31bef87.png)

**Comment by jtkech** [comment]
But if i publish i get a 404 because it doesn't find the specified parent menu

**Comment by mmdqq** [comment]
however i think the problem is MapAreaRoute's template  cannot match more than two parameters,but how two fix it?

**Comment by Skrypt** [comment]
I did not get this 404 let me try again.

**Comment by jtkech** [comment]
Then i tried to add an existing contentItemId as the third parameter, for this i edited the `About` content item

    https://localhost:44300/SubMenu/Create/LinkMenuItem/4zv8qc23wtjgx1f8cgm89ehxmt/4stdftphe3qvyvrrggyc4xmnwz

Then, i got the same above page, and when publishing the new content item has been added, and just under the about item (its parent) and on the left (as a child). Knowing that a content item is a menu with its own items ...

Good luck, need to leave.

**Comment by Skrypt** [comment]
Well let's see if at least the controller action will get hit. I will fake the id's. 
Actually the controller is getting hit but the **menuContentItemId** contains both params : 

4js2vph5exx0p4424dfpcjyg42&menuitemid=4p3jz6ezjhba0vqhsxv7cxtkxf

That's the actual issue here the route mapping is not working properly. And I get the same 404 error.
